### PR TITLE
Add Logging Framework

### PR DESCRIPTION
**Pull request by cwfitzgerald**
This PR adds all the infrastructure needed to log things properly. While I have added 4 log points, that was only for testing, logging is still tbd. 

To be done:
- Translator from bincode to human readable
- Add log points to the code

### CODE REVIEW

File: bve/src/log/global.rs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+/// - `dest` the destination of the logger. Can be a file, stdout/err, or anything else that implements `Write` and
+///   `Send` and `'static`.
+/// - `method` the [format](SerializationMethod) the logger should serialize its messages in.
+///
+/// # Returns
+///
+/// [`GlobalLoggerGuard`] which automatically removes the global logger when it goes out of scope, making sure
+/// everything gets cleaned up correctly.
+pub fn set_global_logger(dest: impl Write + Send + 'static, method: SerializationMethod) -> GlobalLoggerGuard {
+    *GLOBAL_LOGGER.lock().expect("Cannot lock to set global logger") = Some(Subscriber::new(dest, method));
```
Comment (DethRaid) [inline bve/src/log/global.rs]: What is this syntax?

File: .gitignore [hunk 2]
```diff
@@ -7,4 +7,8 @@ cargo-timing*.html
 
 # often need a place to throw one-off files
 /tmp
-/temp
\ No newline at end of file
+/temp
+
+# Log files
+log.*
```
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply .gitignore]: log.txt was my thought

File: bve/src/filesystem/utf8.rs [hunk 3]
```diff
@@ -9,45 +9,60 @@ use std::path::Path;
...
     } else if bytes.len() >= 2 && bytes[0..2] == [0xFE, 0xFF] {
-        encoding_rs::UTF_16BE
+        (encoding_rs::UTF_16BE, "BOM")
     } else if bytes.len() >= 3 && bytes[0..3] == [0xEF, 0xBB, 0xBF] {
-        encoding_rs::UTF_8
+        (encoding_rs::UTF_8, "BOM")
     } else {
+        let span: tracing::Span = tracing::trace_span!("chardetng", size = bytes.len());
+        let _guard = span.enter();
+        tracing::debug!("blah");
```
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply bve/src/filesystem/utf8.rs]: That may or may not have been a testing message that snuck in there......

File: .gitignore (hunk 2 above)
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply .gitignore]: But I should also include that.

File: bve/src/log/method.rs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+pub enum SerializationMethod {
+    Bincode,
+    Json,
+    JsonPretty,
+}
+
+impl SerializationMethod {
+    pub fn serialize(&self, writer: &mut impl Write, message: &Message) {
+        match self {
+            Self::Bincode => bincode::serialize_into(writer, message).expect("Bincode serialization failed"),
```
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply bve/src/log/method.rs]: Json is a visual format, whereas bincode is binary, so doesn't need spaces. This whole thing needs to be reworked to make the json deserializable, but that's a later story.

File: bve/src/log/global.rs (hunk 1 above)
Comment (cwfitzgerald) [reply bve/src/log/global.rs]: Nothing fancy, just assigning a `Option<Subscriber>` to a `MutexLockGuard<Option<Subscriber>>` by deferencing the MutexLockGuard

File: .gitignore (hunk 2 above)
[the remark this answers is not in the document]
Comment (DethRaid) [reply .gitignore]: It's just that `.log` is a common extension for log files. IDK how customizable you need everything to be but I'd ignore `*.log` just to be safe

File: bve/src/log/method.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (DethRaid) [reply bve/src/log/method.rs]: uhm

`writeln!` adds spaces? Or a newline?

That's incredibly not clear from the code. I'd recommend adding the space/newline in a more clear way

Or maybe this is a super common Rust convention that I just don't know about

File: bve/src/log/global.rs (hunk 1 above)
Comment (DethRaid) [reply bve/src/log/global.rs]: I do not like this syntax but that's not your fault

File: bve/src/log/method.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply bve/src/log/method.rs]: It adds a newline. Sorry am extremely tired. It seems common enough that there's a clippy lint about it. https://rust-lang.github.io/rust-clippy/master/index.html#writeln_empty_string

File: bve/src/filesystem/utf8.rs [hunk 5]
```diff
@@ -9,45 +9,60 @@ use std::path::Path;
 ///
 /// Returns Err if opening/reading the file fails. All errors come from [`std::fs::read`].
 pub fn read_convert_utf8(filename: impl AsRef<Path>) -> Result<String> {
+    let span: tracing::Span = tracing::debug_span!("Read to UTF-8", filename = ?filename.as_ref());
```
[the remark this answers is not in the document]
Comment (DethRaid) [reply bve/src/filesystem/utf8.rs]: I did an rtfm and I get it

**Pull request merged**